Fixed #31235 -- Made assertQuerysetEqual() compare querysets directly.

hramezani opened the pull request:
https://code.djangoproject.com/ticket/31235
Continue to https://github.com/django/django/pull/12417

hramezani commented:
It seems the new release of [sphinxcontrib-spelling 6.0.0](https://pypi.org/project/sphinxcontrib-spelling/6.0.0/) has a problem.
I pinned it to `5.4.0` in `tox.ini`.

smithdc1 commented:
Last time we had failing tests due to a dependency @felixxm pinned it in Jenkins and we left the tox file alone. 

I've managed to use `git bisect` to identify the commit that is causing this. 

```
db8d0b3a14de97ac9ecc571f75c1bc2991455b07 is the first bad commit
commit db8d0b3a14de97ac9ecc571f75c1bc2991455b07
Author: Jon Dufresne <*****>
Date:   Thu Oct 15 20:49:21 2020 -0700

    Replace deprecated imp module with importlib

    Fixes warning when running the tests:

        spelling/sphinxcontrib/spelling/filters.py:11: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses

    Per the docs, use importlib.util.find_spec() instead of
    imp.find_module(). See:

    https://docs.python.org/3/library/imp.html#imp.find_module

    For details on importlib.util.find_spec(), see:

    https://docs.python.org/3/library/importlib.html#importlib.util.find_spec

 sphinxcontrib/spelling/filters.py | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)
```

felixxm commented:
Thanks :+1: I will check this tomorrow (CI doesn't use `tox.ini`).

smithdc1 commented:
Hi @jdufresne I see you have been busy raising pull requests against sphinxcontrib-spelling today. _Not sure_ if you have any thoughts on this?

felixxm commented:
I temporarily pinned `sphinxcontrib-spelling != 6.0.0`, see https://github.com/sphinx-contrib/spelling/issues/95.

pope1ni commented:
@hramezani Please can you make sure you credit the original author from #12417 using `Co-authored-by:`.

hramezani commented:
> @hramezani Please can you make sure you credit the original author from #12417 using `Co-authored-by:`.

Great point @pope1ni.

inglesp commented:
> Please can you make sure you credit the original author from #12417 using Co-authored-by:.

Original author here.  Just wanted to say thanks to @hramezani for getting all the tests passing, and to {@pope1ni, @smithdc1, @felixxm}  for getting this over the line.

felixxm commented on tests/aggregation/tests.py:
```diff
@@ -756,7 +756,8 @@ def test_dates_with_aggregation(self):
                 "datetime.date(2008, 1, 1)"
-            ]
+            ],
+            repr,
```
I think we should update assertions, instead of passing `repr`.

hramezani replied:
I changed the assertion to `assertSequenceEqual`

hramezani commented on tests/basic/tests.py:
```diff
@@ -164,10 +164,11 @@ def test_queryset_delete_removes_all_items_in_that_queryset(self):
              "<Article: Article One>",
-             "<Article: Boring article>"]
+             "<Article: Boring article>"],
+            repr,
```
@felixxm  For this case, I can create the objects and put them on an array and then use them on assertion. like:

```diff
--- a/tests/basic/tests.py
+++ b/tests/basic/tests.py
@@ -157,18 +157,20 @@ class ModelTest(TestCase):
         headlines = [
             'An article', 'Article One', 'Amazing article', 'Boring article']
         some_pub_date = datetime(2014, 5, 16, 12, 1)
+        articles = []
         for headline in headlines:
-            Article(headline=headline, pub_date=some_pub_date).save()
+            articles.append(Article.objects.create(headline=headline, pub_date=some_pub_date))
         self.assertQuerysetEqual(
             Article.objects.all().order_by('headline'),
-            ["<Article: Amazing article>",
-             "<Article: An article>",
-             "<Article: Article One>",
-             "<Article: Boring article>"],
-            repr,
+            [
+                articles[2],
+                articles[0],
+                articles[1],
+                articles[3],
+            ],
         )
         Article.objects.filter(headline__startswith='A').delete()
-        self.assertQuerysetEqual(Article.objects.all().order_by('headline'), ["<Article: Boring article>"], repr)
+        self.assertQuerysetEqual(Article.objects.all().order_by('headline'), [articles[3]])
 
     def test_not_equal_and_equal_operators_behave_as_expected_on_instances(self):
         some_pub_date = datetime(2014, 5, 16, 12, 1)

```
What do you think? Do you have any idea?

pope1ni replied:
You can also use `.bulk_create()`.

hramezani replied:
Yes, I tried it but with the `bulk_create()` I got:
`TypeError: Model instances without primary key value are unhashable`

hramezani commented:
I changed the assertion for some test which use `repr`. I will change the rest of them as soon as possible.

jacobtylerwalls commented on tests/test_utils/tests.py:
```diff
@@ -243,19 +244,44 @@ def setUpTestData(cls):
+            Person.objects.all().order_by('name'),
+            [self.p1, self.p2]
+        )
+
```
Why don't we test the deprecation message itself with a `self.assertWarnsMessage()` case?

jacobtylerwalls commented on docs/topics/testing/tools.txt:
```diff
@@ -1683,15 +1683,17 @@ your test suite.
+
+    If ``transform`` is not supplied and ``values`` is a list of string
+    elements, ``values`` is compared to a list produced by applying ``repr()``
+    to each member of ``qs``.
```
Also, since this is going to get a lot of eyes when folks rewrite their tests, how about we clarify the third possibility, which is the new convenience: direct comparison.

I'm not 100% sure of the wording, but something like: "Otherwise, without a ``transform``, the equality of each member of ``qs`` and ``values`` is asserted directly."

jacobtylerwalls reviewed (commented):
Thanks for picking this up 👍

hramezani commented:
@felixxm :
I rebased the PR and changed a lot of assertions and remove `repr` from `assertQuerysetEqual`. 
There are some assertions that are using `repr`. most of them are for models that use more than one field in `__str__`.

like:
* https://github.com/django/django/pull/13557/files#diff-ad6ec549eed244cc16532a3944710bfee97faeba6e3d21fbea58a5cb544254d2R1013
* https://github.com/django/django/pull/13557/files#diff-ad3486daf49a0018f6d9202fbdaf3392dda74e2543decf9bdc43f37d2ebec8b2R221

I think we can keep them as is. what do you think?

felixxm reviewed (commented):
@hramezani Thanks for updates :+1: I added release notes, deprecation annotations, deprecation tests, changed `assertQuerysetEqual()` to `assertSequenceEqual()`/`assertCountEqual()` where appropriate, and pushed small edits.

I'm going to continue my tests edits later today and tomorrow.

felixxm commented:
buildbot

carltongibson commented on django/test/testcases.py:
```diff
@@ -1040,11 +1040,29 @@ def _fixture_teardown(self):
+                # string and the first element of qs is not (which would be the
+                # case if qs is a flattened values_list).
+                warnings.warn(
+                    "In Django 4.1, repr() will not be called on a queryset "
```
`will not automatically be called`?

carltongibson commented on docs/releases/3.2.txt:
```diff
@@ -413,6 +413,9 @@ Tests
   <django.db.transaction.on_commit>` in a list. This allows you to test such
   callbacks without using the slower :class:`.TransactionTestCase`.
 
+* :meth:`.TransactionTestCase.assertQuerysetEqual` now supports querysets
+  comparison.
+
```
Just reading that in isolation, I think, _"Well what did it do before?"_ 

Can we say a little bit more? 

`…now supports direct querysets comparison, rather than requiring comparison against a string representation.`

(Not 100% happy with that.)

pope1ni replied:
Maybe:
```
…now supports direct comparison against another queryset rather than being restricted to comparison against a list of string representations of objects when using the default value for the `transform` argument.
```

carltongibson commented on docs/releases/3.2.txt:
```diff
@@ -606,3 +609,8 @@ Miscellaneous
   due to the now automatic ``AppConfig`` discovery. See :ref:`whats-new-3.2`
   for more details.
+
+* Calling ``repr()`` on a queryset in
```
Automatically calling...

pope1ni replied:
Or maybe we should refer to the fact that the default value of the `transform` argument is changing? Users may not be aware that ``repr()`` is being called for them anyway unless they've played with that argument?

felixxm replied:
It's under the "Features deprecated in 3.2" so an immediate change in the default values doesn't fit here, IMO.

carltongibson commented on docs/topics/testing/tools.txt:
```diff
@@ -1683,15 +1683,17 @@ your test suite.
+    If ``transform`` is provided, ``values`` is compared to a list produced by
+    applying ``transform`` to each member of ``qs``.
+
+    If ``transform`` is not provided and ``values`` is a list of strings, it's
+    compared to a list produced by applying ``repr()`` to each member of
+    ``qs``.
```
This behavior will be removed in Django 4.1. 

I know we don't say that normally... — but can we somehow highlight "Don't use this!" 

Maybe the description of this behaviour can be moved to the `.. deprecated::` block. (Someone beginning to use this method shouldn't use it right? 🤔)

felixxm replied:
Done.

carltongibson commented on tests/aggregation/tests.py:
```diff
@@ -752,11 +752,11 @@ def test_dates_with_aggregation(self):
-                "datetime.date(2007, 1, 1)",
-                "datetime.date(2008, 1, 1)"
-            ]
+                datetime.date(1991, 1, 1),
+                datetime.date(1995, 1, 1),
+                datetime.date(2007, 1, 1),
+                datetime.date(2008, 1, 1),
+            ],
```
This is nice. Worth the change.

carltongibson reviewed (commented):
Added a few thoughts for `django/` and `docs/`.

pope1ni commented on docs/internals/deprecation.txt:
````diff
@@ -26,6 +26,9 @@ details on these changes.
 
 * The ``default_app_config`` module variable will be removed.
 
+* ``TransactionTestCase.assertQuerysetEqual()` will no longer call ```repr()``
````
```suggestion
* ``TransactionTestCase.assertQuerysetEqual()` will no longer call ``repr()``
```

pope1ni commented on docs/topics/testing/tools.txt:
```diff
@@ -1706,7 +1706,17 @@ your test suite.
+
+    .. versionadded:: 3.2
+
+        Support for querysets comparison was added.
```
```suggestion
        Support for direct comparison between querysets was added.
```

Strictly speaking this was always possible if you passed `transform=lambda x: x`, but was just not obvious. So does this really count as `versionadded` or just another part of `versionchanged`? Maybe we should say that the value was changed and this is now what it means.

felixxm replied:
I think we can leave `versionadded`.

felixxm commented:
@carltongibson @pope1ni Thanks for reviews. I pushed edits to docs. I'm still working on tests...

felixxm commented:
Pushed more tests edits, I'm going to finish tomorrow :moon:

felixxm commented:
buildbot, test on oracle.

felixxm commented:
OK, I finally finished my edits.

carltongibson reviewed (approved):
Thanks for updating @felixxm. I think the docs changes are clearer. I'd say move forward. 👍 

Thanks, as always, for picking this up @hramezani! 🥇

felixxm commented:
Thanks y'all :rocket:

felixxm commented:
buildbot, test on oracle.
